File: sdk-python/dephealth/endpoint_status.py
"""EndpointStatus: detailed health check state for a single endpoint."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime


@dataclass(frozen=True)
class EndpointStatus:
    """Detailed health check state for a single endpoint.

    Returned by health_details(). Contains all 11 fields defined in spec section 8.
    """

    healthy: bool | None
    status: str
    detail: str
    latency: float
    type: str
    name: str
    host: str
    port: str
    critical: bool
    last_checked_at: datetime | None
    labels: dict[str, str] = field(default_factory=dict)

    def latency_millis(self) -> float:
        """Return the latency in milliseconds."""
        return self.latency * 1000.0

    def to_dict(self) -> dict[str, object]:
        """Serialize to a JSON-compatible dictionary.

        Latency is serialized as ``latency_ms`` (milliseconds float).
        LastCheckedAt is serialized as ISO 8601 string or None.
        """
        return {
            "healthy": self.healthy,
            "status": self.status,
            "detail": self.detail,
            "latency_ms": self.latency_millis(),
            "type": self.type,
            "name": self.name,
            "host": self.host,
            "port": self.port,
            "critical": self.critical,
            "last_checked_at": (self.last_checked_at.isoformat() if self.last_checked_at else None),
            "labels": dict(self.labels),
        }


__all__ = [
    "EndpointStatus",
]

File: sdk-go/dephealth/endpoint_status.go
package dephealth

import (
	"encoding/json"
	"time"
)

// EndpointStatus represents the detailed health check state for a single endpoint.
// It is returned by HealthDetails() and contains all 11 fields defined in the specification.
type EndpointStatus struct {
	Healthy       *bool             `json:"healthy"`
	Status        StatusCategory    `json:"status"`
	Detail        string            `json:"detail"`
	Latency       time.Duration     `json:"-"`
	Type          DependencyType    `json:"type"`
	Name          string            `json:"name"`
	Host          string            `json:"host"`
	Port          string            `json:"port"`
	Critical      bool              `json:"critical"`
	LastCheckedAt time.Time         `json:"last_checked_at"`
	Labels        map[string]string `json:"labels"`
}

// LatencyMillis returns the latency in milliseconds as a float64.
func (es EndpointStatus) LatencyMillis() float64 {
	return float64(es.Latency.Nanoseconds()) / 1e6
}

// endpointStatusJSON is the JSON representation of EndpointStatus.
type endpointStatusJSON struct {
	Healthy       *bool             `json:"healthy"`
	Status        StatusCategory    `json:"status"`
	Detail        string            `json:"detail"`
	LatencyMs     float64           `json:"latency_ms"`
	Type          DependencyType    `json:"type"`
	Name          string            `json:"name"`
	Host          string            `json:"host"`
	Port          string            `json:"port"`
	Critical      bool              `json:"critical"`
	LastCheckedAt *time.Time        `json:"last_checked_at"`
	Labels        map[string]string `json:"labels"`
}

// MarshalJSON implements custom JSON marshaling.
// Latency is serialized as latency_ms (milliseconds float).
// LastCheckedAt is serialized as null when zero (before first check).
func (es EndpointStatus) MarshalJSON() ([]byte, error) {
	j := endpointStatusJSON{
		Healthy:   es.Healthy,
		Status:    es.Status,
		Detail:    es.Detail,
		LatencyMs: es.LatencyMillis(),
		Type:      es.Type,
		Name:      es.Name,
		Host:      es.Host,
		Port:      es.Port,
		Critical:  es.Critical,
		Labels:    es.Labels,
	}
	if !es.LastCheckedAt.IsZero() {
		t := es.LastCheckedAt.UTC()
		j.LastCheckedAt = &t
	}
	return json.Marshal(j)
}

// UnmarshalJSON implements custom JSON unmarshaling.
func (es *EndpointStatus) UnmarshalJSON(data []byte) error {
	var j endpointStatusJSON
	if err := json.Unmarshal(data, &j); err != nil {
		return err
	}
	es.Healthy = j.Healthy
	es.Status = j.Status
	es.Detail = j.Detail
	es.Latency = time.Duration(j.LatencyMs * 1e6)
	es.Type = j.Type
	es.Name = j.Name
	es.Host = j.Host
	es.Port = j.Port
	es.Critical = j.Critical
	es.Labels = j.Labels
	if j.LastCheckedAt != nil {
		es.LastCheckedAt = *j.LastCheckedAt
	}
	return nil
}
